deploy/v2beta1/mpi-operator.yaml

```diff
@@ -30,233 +33,8018 @@ spec:
-                    type: object
-                    x-kubernetes-preserve-unknown-fields: true
-                type: object
-              slotsPerWorker:
-                minimum: 1
-                type: integer
-            type: object
-            x-kubernetes-preserve-unknown-fields: true
-          status:
-            type: object
-            x-kubernetes-preserve-unknown-fields: true
-        type: object
-    served: true
-    storage: false
-    subresources:
-      status: {}
   - name: v2beta1
     schema:
       openAPIV3Schema:
         properties:
+          apiVersion:
+            description: 'APIVersion defines the versioned schema of this representation
+              of an object. Servers should convert recognized schemas to the latest
+              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+            type: string
+          kind:
+            description: 'Kind is a string value representing the REST resource this
+              object represents. Servers may infer this from the endpoint the client
+              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
+            type: string
+          metadata:
+            type: object
           spec:
             properties:
               mpiImplementation:
+                default: OpenMPI
+                description: MPIImplementation is the MPI implementation. Options
+                  are "OpenMPI" (default) and "Intel".
                 enum:
                 - OpenMPI
                 - Intel
                 type: string
               mpiReplicaSpecs:
-                properties:
-                  Launcher:
-                    properties:
-                      replicas:
-                        maximum: 1
-                        minimum: 1
-                        type: integer
-                      restartPolicy:
-                        enum:
-                        - Never
-                        - OnFailure
-                        type: string
-                      template:
-                        type: object
-                        x-kubernetes-preserve-unknown-fields: true
-                    type: object
-                  Worker:
-                    properties:
-                      replicas:
-                        minimum: 1
-                        type: integer
-                      restartPolicy:
-                        enum:
-                        - Never
-                        - OnFailure
-                        type: string
-                      template:
-                        type: object
-                        x-kubernetes-preserve-unknown-fields: true
-                    type: object
-                required:
-                - Launcher
+                additionalProperties:
```

Comment:
We are losing the knowledge about Launcher and Worker here.
But this is a problem with the API structs themselves.

We shouldn't be using a map https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Maybe we can fix it in a v2(beta2) API, but it would be nice to fix other training job objects, like so:

```golang
type MPIReplicaSpecs struct {
  Launcher common.ReplicaSpec `json:"launcher"`
  Worker common.ReplicaSpec `json:"worker"`
}
```

Reply:
Looks good.
IIRC, we plan to remove dependencies with kubeflow/common in the future altogether.
So we will be able to fix this after that.

Either way, I create an issue to keep tracking this.